cliapp/example2.1.in
.\" Copyright (C) 2011  Lars Wirzenius
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH EXAMPLE2.PY 1
.SH NAME
example2.py \- cliapp-based simplistic hello, world app
.SH SYNOPSIS
.SH DESCRIPTION
.B example2.py
is an example program for
.BR cliapp ,
a Python framework for command line utilities.
.SH OPTIONS
.SH EXAMPLE
To greet:
.IP
python example2.py greet
.PP
To insult:
.IP
python example2.py insult


cliapp/NEWS
NEWS for cliapp
===============

Version 1.20180812.1+git, not yet released
------------------------------------------


Version 1.20180812.1, released 2018-08-12
----------------------------------------


Version 1.20180812, released 2018-08-12
----------------------------------------

* Fix file conflict between python3 and old versions of python2
  versions of the Debian package.

Version 1.20180121, released 2018-01-21
----------------------------------------

* The `runcmd` timeout behaviour has changed a little. Previously, a
  timeout meant the command would be terminated. Termination is now
  optional: if the `timeout_callback` function returns a true value,
  termination happens. Otherwise not.

Version 1.20170827, released 2017-08-27
----------------------------------------

* Fix settings to not use Unicode strings (drop import of Unicode
  literals from the future).

Version 1.20170823, released 2017-08-23
----------------------------------------

* Only Debian packaging changes.

Version 1.20170820, released 2017-08-20
----------------------------------------

* `cliapp.runcmd` now has a optional arguments `output_timeout` and
  `timeout_callback`.
* Meliae the Python memory profiler is no longer supported.
* Sphinx-generated documentation is no longer supported. It was
  unmainained and rubbish by now.
* Added support for Python 3, thanks to patches from Stuart Prescott.

Version 1.20160724, released 2016-07-24
----------------------------------------

* Debian packaging change: running automated checks during package
  build.

Version 1.20160316, released 2016-03-16
---------------------------------------

* The `--help` now shows the previously hidden options
  `--dump-config-names`, `--list-config-files`, and
  `--generate-manpage`. These are not frequently used, but sometimes
  they are useful, and there's no point making them difficult to find.

* The **cliapp**(5) manual page now documents the units for options
  whose value is a size in bytes.

Version 1.20151108, released 2016-01-09
---------------------------------------

Bug fix:

* The generated manpages used to have an equals sign for short options
  with an argument. For example, `-e=REGEX`. This has now been fixed
  to show the correct form, `-eREGEX`, instead.

* If a user specifies a config file with `--config`, it must exist.

* The automatically added negated boolean options (`--no-foo` added
  for `--foo`) are now sorted after the non-negated ones in manual
  pages.

New features:

* Configuration files may now also be in YAML format, if names with an
  `.yaml` suffix. This also means that the `python-yaml` library
  **MUST** installed. It is not an optional dependency.

* If the `python-xdg` library is available, it is used to allow user
  to specify locations of XDG Base Directory locations.

* Log to STDERR support has been added. Use `--log=stderr`. Patch by
  Roland Mas.

Version 1.20150829
------------------

Important, backwards incompatible bug fixes:

* Jan Gerber fixed string list option handling. Command line options
  no longer parse values by commas, so that `--foo=bar,foobar` results
  in a single value `bar,foobar`, instead of two values (`bar` and
  `foobar`). Also, in configuration files, values may be quoted with
  double quotes.

Bug fixes:

* Memory use profiling with Meliae was fixed.

New features:

* Richard Ipsum added the `cliapp.Application.get_subcommand_usage`
  method to return a short usage text for a subcommand. This allows
  better error message.

Version 1.20150701
------------------

* Lars Wirzenius added the `ssh_options` keyword argument to
  `cliapp.ssh_runcmd` to allow adding command line options to the ssh
  program.

* Lars Wirzenius added the `remote_cwd` keyword argument to
  `cliapp.ssh_runcmd` to allow the caller to easily change to a
  directory on the server.

* Support for `heapy` memory profile has been dropped. It's not
  packaged for Debian, and is rarely available anywhere.


Version 1.20150305
------------------

* Richard Ipsum added callbacks to `cliapp.runcmd` for handling
  captured stdout/stderr output from the pipeline. This allows, for
  example, progress reporting during a long-running command.

* Richard Maw and Richard Ipsum fixed `cliapp.runcmd` to close file
  descriptors correctly, to make `cliapp.runcmd(['cat', '/dev/zero'],
  ['false'])` work correctly. Previously, the pipeline would never
  finish, but now it does, because closing an unnecessary file
  descriptor in the parent results in `SIGPIPE` being sent to the
  `cat` process once `false` terminates.

* Richard Maw changed `cliapp.runcmd` so that the stderr of all
  commands in the pipeline are collected, when stderr is set to
  `subprocess.PIPE`.

Version 1.20140719
------------------

* The way logging is set up has been split into smaller methods, to
  allow overriding better. See `setup_logging_handler_for_file`,
  `setup_logging_formatter_for_file`, and
  `setup_logging_formatter_for_syslog`.

* Plugins no longer need to define a `disable` method: the default
  implementation is now a no-op, instead of raising
  `NotImplementedError`.

Bug fixes:

* When getting help for a subcommand, cliapp would crash saying
  `get_help_text_formatter` couldn't be found. This has been fixed.

Version 1.20140315
------------------

* Portability patch to disable VmRSS reporting on non-Linux platforms.
  The code assumed that /proc/self/status exists, but that's only true
  on Linux. Patch from Itamar Turner-Trauring.

* `cliapp` now logs the current working directory, uid, effective uid,
  gid, and effective gid at startup.

* `cliapp` (`Settings.load_configs`) now reports an unknown
  variable in a configuration file with a nice error message, rather
  than a stack trace.

* A new method, `cliapp.Application.get_subcommand_help_formatter`
  allows overriding how the full help text for a subcommand is to be
  formatted. This can be useful for allowing help texts be marked up

cliapp/example2.py
# Copyright (C) 2011  Lars Wirzenius
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


'''Example for cliapp framework.

Greet or insult people.

'''


import cliapp


class ExampleApp(cliapp.Application):

    cmd_synopsis = {
        'greet': 'USER...',
        'insult': 'USER...',
    }

    def cmd_greet(self, args):
        '''Greet the user.

        The user is treated to a a courteus,
        but terse form of greeting.

        '''
        if len(args) == 0:
            raise cliapp.AppException(self.get_subcommand_usage('greet'))

        for arg in args:
            self.output.write('greetings, %s\n' % arg)

    def cmd_insult(self, args):
        '''Insult the user.

        Sometimes, though rarely, it happens that a user is really a bit of
        a prat, and needs to be told off. This is the command for that.

        '''
        if len(args) == 0:
            raise cliapp.AppException(self.get_subcommand_usage('insult'))

        for arg in args:
            self.output.write('you suck, %s\n' % arg)


app = ExampleApp(
    version='0.1.2',
    description='''
Greet the user.
Or possibly insult them. User's choice.
''',
    epilog='''
This is the epilog.

I hope you like it.
''')

app.run()
